### ISSUE TITLE
[JENKINS-37340] Need another pubsub event to indicate that a job run has actually started doing some work

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#3117]
<p>From <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a><br/>
The server side events for Pipeline jobs fire immediately. If the user presses the "Build Now" from Classic UI eight times, the Blue Ocean UI will receive events showing status of "QUEUED", "ALLOCATED" and "RUNNING" for all eight jobs immediately. Meanwhile the classic UI shows six of the jobs as queued, while two are running (what the user would expect with only two executors active.</p>

<p>Some implementation ideas, from <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a><br/>
Introduce custom logic in "pubsub-light" SyncRunListener to not fire start events for Pipeline jobs. Create some kind of new listener - "PipelineSyncRunListener, possibly in a new module like pipeline-pubsub-light" - when receives notification of Pipeline's step/stage start and then produces the proper start event as in SyncRunListener.</p>

<p>and impl ideas from <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly" class="user-hover" rel="tfennelly">tfennelly</a><br/>
Keep the existing "started" event but add a new event called "running" which the Blue Ocean UI would use to transition the job from "queued" to "running"</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cliffmeyers.png" title="cliffmeyers's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers">cliffmeyers</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37340" target="_blank">Need another pubsub event to indicate that a job run has actually started doing some work</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly">tfennelly</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Critical
<li><b>component(s)</b>: blueocean-plugin, sse-gateway-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2016-10-18T00:38:03+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 3
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>From <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a>
The server side events for Pipeline jobs fire immediately. If the user presses the "Build Now" from Classic UI eight times, the Blue Ocean UI will receive events showing status of "QUEUED", "ALLOCATED" and "RUNNING" for all eight jobs immediately. Meanwhile the classic UI shows six of the jobs as queued, while two are running (what the user would expect with only two executors active.</p>

<p>Some implementation ideas, from <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a>
Introduce custom logic in "pubsub-light" SyncRunListener to not fire start events for Pipeline jobs. Create some kind of new listener - "PipelineSyncRunListener, possibly in a new module like pipeline-pubsub-light" - when receives notification of Pipeline's step/stage start and then produces the proper start event as in SyncRunListener.</p>

<p>and impl ideas from <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly" class="user-hover" rel="tfennelly">tfennelly</a>
Keep the existing "started" event but add a new event called "running" which the Blue Ocean UI would use to transition the job from "queued" to "running"</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> tfennelly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37340?focusedId=266837&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-266837">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The run_started event fires at the right time i.e. when the run moves out of the queue. The problem with pipeline runs is that this happens immediately as the run creates a "lightweight" run task that's not really doing anything yet. We need another pubsub event to let listeners know when the run has actually started "running" some of the pipeline steps.</p>

<p>Please remember that the pubsub stuff should remain generic i.e. not specific to the UI needs, so when we hit things like this we need to address them in a general way Vs just having the UI goggles on.</p>
</pre>
</details>

</ul>
</details>

<p>The run_started event fires at the right time i.e. when the run moves out of the queue. The problem with pipeline runs is that this happens immediately as the run creates a "lightweight" run task that's not really doing anything yet. We need another pubsub event to let listeners know when the run has actually started "running" some of the pipeline steps.</p>

<p>Please remember that the pubsub stuff should remain generic i.e. not specific to the UI needs, so when we hit things like this we need to address them in a general way Vs just having the UI goggles on.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37340?focusedId=266838&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-266838">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a> is this blocking favourite realtime stuff? </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a> is this blocking favourite realtime stuff? </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37340?focusedId=267596&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-267596">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Should really see what cliff cooks up on <a href="https://issues.jenkins.io/browse/JENKINS-37379" title="Some real-time favorite card statuses are not accurate" class="issue-link" data-issue-key="JENKINS-37379">JENKINS-37379</a> - before tackling this. May not be needed.</p>
</pre>
</details>

</ul>
</details>

<p>Should really see what cliff cooks up on <a href="https://issues.jenkins.io/browse/JENKINS-37379" title="Some real-time favorite card statuses are not accurate" class="issue-link" data-issue-key="JENKINS-37379">JENKINS-37379</a> - before tackling this. May not be needed.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cliffmeyers.png" title="cliffmeyers's avatar" /> cliffmeyers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37340?focusedId=267787&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-267787">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly" class="user-hover" rel="tfennelly">tfennelly</a>: yep, let me see if I can unblock it myself, rather than having to make the pub/sub stuff more involved. Will let you know once I get to <a href="https://issues.jenkins.io/browse/JENKINS-37379" title="Some real-time favorite card statuses are not accurate" class="issue-link" data-issue-key="JENKINS-37379">JENKINS-37379</a>, should be later this week.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly" class="user-hover" rel="tfennelly">tfennelly</a>: yep, let me see if I can unblock it myself, rather than having to make the pub/sub stuff more involved. Will let you know once I get to <a href="https://issues.jenkins.io/browse/JENKINS-37379" title="Some real-time favorite card statuses are not accurate" class="issue-link" data-issue-key="JENKINS-37379">JENKINS-37379</a>, should be later this week.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37340?focusedId=273724&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-273724">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a> is this something you still want tom to take a look at ?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a> is this something you still want tom to take a look at ?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cliffmeyers.png" title="cliffmeyers's avatar" /> cliffmeyers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37340?focusedId=273776&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-273776">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>No, I believe that once <a href="https://issues.jenkins.io/browse/JENKINS-38540" title="Add REST API redirects for /runs and /queue" class="issue-link" data-issue-key="JENKINS-38540"><del>JENKINS-38540</del></a> and <a href="https://issues.jenkins.io/browse/JENKINS-37379" title="Some real-time favorite card statuses are not accurate" class="issue-link" data-issue-key="JENKINS-37379">JENKINS-37379</a> are implemented that this issue will be unnecessary.</p>
</pre>
</details>

</ul>
</details>

<p>No, I believe that once <a href="https://issues.jenkins.io/browse/JENKINS-38540" title="Add REST API redirects for /runs and /queue" class="issue-link" data-issue-key="JENKINS-38540"><del>JENKINS-38540</del></a> and <a href="https://issues.jenkins.io/browse/JENKINS-37379" title="Some real-time favorite card statuses are not accurate" class="issue-link" data-issue-key="JENKINS-37379">JENKINS-37379</a> are implemented that this issue will be unnecessary.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> tfennelly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37340?focusedId=273791&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-273791">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I can't see how these fix the underlying issue, but if it saves me doing something then <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>I can't see how these fix the underlying issue, but if it saves me doing something then <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37340?focusedId=273875&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-273875">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>lets close this until we know it is needed. </p>
</pre>
</details>

</ul>
</details>

<p>lets close this until we know it is needed. </p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `epic` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-35737%22">JENKINS-35737</a>]</i>
